src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala

```diff
@@ -106,24 +137,26 @@ private[mock] class NativeMockKafkaProducer(
     record =>
       if (record.topic === topic) keyDeserializer.deserialize(topic, Headers.empty, record.key).map(_ === key)
       else IO.pure(false)
-  ).map(_.map { case (_, _, value) => value })
+  ).map(_.collect { case (_, _, Some(value)) => value })
 
   private def selectedHistory[K, V](
       topic: String,
       recordSelector: ProducerRecord[Array[Byte], Array[Byte]] => IO[Boolean]
   )(implicit
       keyDeserializer: KeyDeserializer[IO, K],
       valueDeserializer: ValueDeserializer[IO, V]
-  ): IO[List[(Int, K, V)]] = mockProducer.history.asScala.zipWithIndex.toList
+  ): IO[List[(Int, K, Option[V])]] = mockProducer.history.asScala.zipWithIndex.toList
     .traverse { case (record, index) =>
       for {
         isSelected <- recordSelector(record)
         key <-
           if (isSelected) keyDeserializer.deserialize(topic, Headers.empty, record.key).map(Option(_))
           else IO(None)
         value <-
-          if (isSelected) valueDeserializer.deserialize(topic, Headers.empty, record.value).map(Option(_))
-          else IO(None)
+          (if (isSelected) {
+             if (record.value eq null) IO(Some(None))
```

Comment:
wondering if using `Option(value)` would have done the trick also?

Reply:
Yup, I tried that first, but it was kind of tricky to the point that I felt it would be hard to follow, so I decided to make it explicit (i.e. I gave up ;-))

Reply:
Consider me nerdsniped:
```
if (isSelected) Option(record.value).traverse(valueDeserializer.deserialize(topic, Headers.empty, _)).map(Some(_))
else IO(None)
```

I think this is equivalent, but I had to think a lot harder :-) Up to you what you prefer.

Reply:
glad i could get you to rev up the engine 😄  ⚙️ I like the second one better I think, nested ifs and 'null' kind of make me restless nowadays 😆